### ISSUE TITLE
client.send_produce_request with fail_on_error=False actually will fail due to LeaderNotAvailableError

### ISSUE THREAD

**Issue by reAsOn2010** [dpkp/kafka-python#414]
In async producer we use a thread to send message. We call `client.send_produce_request` with `fail_on_error=False` as parameter. But if a partition really has no leader now, a LeaderNotAvailableError is raised **in thread**. I don't know is it desirable, but this will let the thread crash and it seems that there is no easy way to restart such a thread. If we continue use this producer, no message will be send.

Below is the stack

```
[E 150618 16:39:01 /Users/leon/opensource/kafka-python/kafka/conn.py:106 leon-MacBook-Pro:17631] Unable to receive data from Kafka
Traceback (most recent call last):
  File "/Users/leon/opensource/kafka-python/kafka/conn.py", line 103, in _read_bytes
    raise socket.error("Not enough data to read message -- did server kill socket?")
error: Not enough data to read message -- did server kill socket?
[W 150618 16:39:01 /Users/leon/opensource/kafka-python/kafka/client.py:215 leon-MacBook-Pro:17631] ConnectionError attempting to receive a response to request 15 from server BrokerMetadata(nodeId=2, host='kids.aws.dev', port=9094): Kafka @ kids.aws.dev:9094 went away
[E 150618 16:39:01 /Users/leon/opensource/kafka-python/kafka/producer/base.py:174 leon-MacBook-Pro:17631] FailedPayloadsError sending ProduceRequest (#1 of 1) to kafka-test:4 with msgs (Message(magic=0, attributes=0, key=None, value='[I 150618 16:39:01 /Users/leon/zhihu/timeline/timeline/api.py:44 leon-MacBook-Pro:17631] ping id: 12'),)
[W 150618 16:39:01 /Users/leon/opensource/kafka-python/kafka/producer/base.py:182 leon-MacBook-Pro:17631] Async producer backoff for 100(ms) before retrying
[I 150618 16:39:01 /Users/leon/opensource/kafka-python/kafka/producer/base.py:204 leon-MacBook-Pro:17631] Retrying ProduceRequest to kafka-test:4 with msgs (Message(magic=0, attributes=0, key=None, value='[I 150618 16:39:01 /Users/leon/zhihu/timeline/timeline/api.py:44 leon-MacBook-Pro:17631] ping id: 12'),)
[I 150618 16:39:01 /Users/leon/opensource/kafka-python/kafka/client.py:366 leon-MacBook-Pro:17631] Updating broker metadata: [BrokerMetadata(nodeId=0, host='kids.aws.dev', port=9092), BrokerMetadata(nodeId=1, host='kids.aws.dev', port=9093), BrokerMetadata(nodeId=3, host='kids.aws.dev', port=9095)]
[I 150618 16:39:01 /Users/leon/opensource/kafka-python/kafka/client.py:367 leon-MacBook-Pro:17631] Updating topic metadata: [TopicMetadata(topic='kafka-test', error=0, partitions=[PartitionMetadata(topic='kafka-test', partition=2, leader=0, replicas=(0,), isr=(0,), error=0), PartitionMetadata(topic='kafka-test', partition=5, leader=3, replicas=(3,), isr=(3,), error=0), PartitionMetadata(topic='kafka-test', partition=4, leader=-1, replicas=(), isr=(), error=5), PartitionMetadata(topic='kafka-test', partition=1, leader=3, replicas=(3,), isr=(3,), error=0), PartitionMetadata(topic='kafka-test', partition=3, leader=1, replicas=(1,), isr=(1,), error=0), PartitionMetadata(topic='kafka-test', partition=0, leader=-1, replicas=(), isr=(), error=5)])]
[E 150618 16:39:01 /Users/leon/opensource/kafka-python/kafka/client.py:405 leon-MacBook-Pro:17631] No leader for topic kafka-test partition 4
[E 150618 16:39:01 /Users/leon/opensource/kafka-python/kafka/client.py:405 leon-MacBook-Pro:17631] No leader for topic kafka-test partition 0
Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/Users/leon/opensource/kafka-python/kafka/producer/base.py", line 154, in _send_upstream
    fail_on_error=False)
  File "/Users/leon/opensource/kafka-python/kafka/client.py", line 476, in send_produce_request
    resps = self._send_broker_aware_request(payloads, encoder, decoder)
  File "/Users/leon/opensource/kafka-python/kafka/client.py", line 166, in _send_broker_aware_request
    payload.partition)
  File "/Users/leon/opensource/kafka-python/kafka/client.py", line 95, in _get_leader_for_partition
    raise LeaderNotAvailableError(meta)
LeaderNotAvailableError: PartitionMetadata(topic='kafka-test', partition=4, leader=-1, replicas=(), isr=(), error=5)
```

From then on, the producer send no messages.

Looking forward to responses~ THX~ 😄

**Comment by mre** [comment]
Same here.
As a workaround, you could try catching the exception like so:

```Python
try:
    resps = kafka.send_produce_request(payloads=[req])
except LeaderNotAvailableError:
   time.sleep(1)
   resps = kafka.send_produce_request(payloads=[req])
```

See also: https://github.com/dpkp/kafka-python/issues/249

**Comment by dpkp** [comment]
fixed by PR #467
